doc/mgmt/gnmi/SONiC_GNMI_Server_Interface_Design.md

```diff
@@ -0,0 +1,530 @@
+| gNOI     | gRPC Network Operations Interface |
+| JSON     | JavaScript Object Notation |
+| Yang     | Yet Another Next Generation, it is a modular language representing data structures in an XML tree format |
+| Generic Config Updater | SONiC configuration generic update and rollback feature |
+| ConfigDB | Configuration Database |
+| ApplDB | Application Database |
+
+# 1 Feature Overview
+
+SONiC should provide a gNMI/gNOI server interface where the client can communicate with SONiC (server) through the interface using their respective gRPC libraries based on ConfigDB/ApplDB/StateDB/CountersDB schema or SONiC Yang schema.
+
+To understand the requirements, we need to go over current SONiC management components:
+
+* SONIC CLI commands
+
+<img src="images/cli.png" alt="cli" width="500px"/>
+
+1. Users could use SONiC CLI command "show runningconfig all" to read current configuration. 
+    - The blue arrow indicates this process.
+2. User could use SONiC CLI command "config apply-patch" to incrementally update configurations systematically and validate target configuration using SONiC Yang models. With generic_config_updater, all operations in a 'config apply-patch' command are processed in a single transaction that will either succeed or fail as a whole.
+    - "config apply-patch" supports ConfigDB, and Yang models.
+    - The green arrow indicates this process, and dotted arrow indicates Yang validation.
+3. SONiC CLI commands can use "config reload" to fully update configurations and validate target configuration using SONiC Yang models.
+    - "config reload" already support input in ConfigDB schema or in yang models schema.
+    - The yellow arrow indicates this process, and dotted arrow indicates Yang validation.
+
+And we have the possibility to improve generic_config_updater performance.
+
+* SONIC Restapi
+
+<img src="images/restapi.png" alt="restapi" width="500px"/>
+
+sonic-restapi is a docker container running in SONiC, which exposes HTTPS endpoints to perform dynamic configuration including config VNET routes.
+One challenge of sonic-restapi is to provide a general config schema. Currently its API is designed case by case.
+
+sonic-restapi will be replaced by gNMI interface in the future.
+
+* SONiC system telemetry services
+
+<img src="images/telemetry.png" alt="telemetry" width="500px"/>
+
+sonic-telemetry provides gNMI/gNOI server interface. And sonic-telemetry provides only telemetry via gRPC, and the gap is configuration.
+
+## 1.1 Requirements
+
+### 1.1.1 Phase 1 Requirements
+* Set and get RPCs must be supported. Customer will use get RPC to retrieve configurations including FRR and VNET route, and use set RPC to apply new configurations.
+* SetRequest message must support to incrementally update configuration and fully update configurations.
+* Data models can be SONiC Yang models or CONFIG_DB schema.
+* Ability to configure/read different DBs - ApplDB, ConfigDB, StateDB, CountersDB etc.
```

Comment:
Please see the comment above, writing directly to APPL_DB/STATE_DB/COUNTERS_DB is against the general norm for SONIC management model.